### DISCUSSION TITLE
What's the correct way to get FIPS implementations?

### DISCUSSION THREAD

**Discussion by kulkarniamit** [openssl/openssl#26170] [Q&A] [answered]
#### Approach 1
* Load fips provider using `OSSL_PROVIDER_try_load(NULL, "fips", 1)` with `retain_fallbacks` set to `1` to retain the fallback provider (`default`)
* Enable `fips=yes` default property for all operations by calling `EVP_default_properties_enable_fips(NULL, 1)` to ensure we fetch implementations that have `fips=yes` property set.
* Encoders and decoders will be fetched from `default` provider since it is enabled as fallback ([OSSL_PROVIDER-default](https://docs.openssl.org/3.0/man7/OSSL_PROVIDER-default/#asymmetric-key-encoder)). Hence no need to load `base` provider.

#### Approach 2
* Load fips provider using `OSSL_PROVIDER_load(NULL, "fips")` before loading any other provider (This disables fallback providers and `default` provider is not loaded)
* Load base provider using `OSSL_PROVIDER_load(NULL, "base")`
* Encoders and decoders will be fetched from `base` provider ([OSSL_PROVIDER-base](https://docs.openssl.org/3.0/man7/OSSL_PROVIDER-base/#asymmetric-key-encoder))

Are both approaches correct or is one of them preferred over the other to ensure all cryptographic operations use FIPS implementations?

**Comment by paulidale** [comment, marked as answer]
Both are correct.  Approach 2 is less prone to error.

**Reply by kulkarniamit** [reply]
How could Approach 1 run into errors if `fips=yes` property applies to all fetches and ensures all returned implementations have the property set on them?

**Reply by paulidale** [reply]
It wouldn't.  But failing to include fips=yes or accidentally setting a different property query on a call could result in non-FIPS validated algorithms being used.  With approach 2, this simply cannot happen.

Given the option, FIPS + base is the way to go to ensure FIPS compliance.  It is easier to justify when you are audited -- simply because other algorithms just aren't available.

You did ask for the preferred option after all.

**Reply by kulkarniamit** [reply]
Got it, thank you for the explanation. That makes sense.

**Comment by StephenWall** [comment]
You can also rely on e.g. /etc/ssl/openssl.conf to force all software using openssl to use only FIPS and base providers.

**Reply by kulkarniamit** [reply]
You're right, that would help applications that use system OpenSSL.
